### PR TITLE
Add episode 12

### PR DESCRIPTION
**Pull request by bitsondatadev**

### CODE REVIEW

File: _episodes/12.md [hunk 1]
```diff
@@ -0,0 +1,263 @@
+---
+layout: episode
+title:  "12: Trino gets super visual with Superset!"
```
Comment (mosabua) [inline _episodes/12.md:3]: Apache Superset

Maybe

File: _episodes/12.md [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+date: 2021-03-04
+tags: trino visualization superset apache client jdbc python
+youtube_id: "0NHUs-TERtk"
+wistia_id: "ggayk0tihh"
+sections: 
+   - title: "Concept of the week"
+     desc: "Trino client, Python, and Apache Superset "
```
Comment (mosabua) [inline _episodes/12.md:10]: Trino clients,

File: _episodes/12.md [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+     time: 463
+   - title: "PR of the week"
+     desc: "Superset PR 13105 feat: first step native support Trino"
+     time: 1076
+   - title: "PR Demo"
```
Comment (mosabua) [inline _episodes/12.md:15]: I think just 

Demo 

might be better

File: _episodes/12.md [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+     desc: "PR Demo: Superset PR 13105 feat: first step native support Trino"
```
Comment (mosabua) [inline _episodes/12.md:16]: Demonstration of Apache Superset querying Trino to create a number of visualizations

Comment (mosabua) [reply _episodes/12.md:16]: Also in section header below..

File: _episodes/12.md [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+   client does the same more directly
+ * PyODBC 
+   * Similar hack to PyJDBC
+ * Potentially also possible to talk to via HTTP directly
+   * That’s like reimplementing the trino-python-client
+   * Also see question of the week later
+
+Beyond that, it will vary from application to application.
+
+Let’s find out from our guests how this hangs together in Apache Superset, since
```
Comment (mosabua) [inline _episodes/12.md:105]: We need an addition section for the TCB that has something like

Live guests

Jane Done from someorg, twitter url, github url

or something like that

We should at least add our guest from this episode on the text

File: _episodes/12.md [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+
+In this week's pull request <https://github.com/apache/superset/pull/13105> that
+was graciously added by [dungdm93](https://github.com/dungdm93).
+
+The first thing we need to understand about this addition is the concept of a
+database engine in Superset. A database engine handles a lot of the custom
+interactions between various databases and maps them to the interface that 
+Superset understands. If certain concepts are missing in a certain database, 
+like time granularity or SQL syntax, the database engine for that database
+will indicate to Superset that this is not available, and it won't show up as an
```
Comment (mosabua) [inline _episodes/12.md:118]: that database indicated to Superset that this is not available. As a result the option does not show in Superset or a concise error message is reported.

File: _episodes/12.md [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+option or report a concise error to the user. By default, database engines use
+the [base.py](https://github.com/apache/superset/blob/master/superset/db_engine_specs/base.py)
+methods, but each engine, like Trino, will add the custom mappings via the
```
Comment (mosabua) [inline _episodes/12.md:121]: like Trino, add custom mappings with a specific engine implementeation (trino.py).

Forget about the verify stuff.

File: _episodes/12.md [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+[trino.py engine](https://github.com/apache/superset/blob/master/superset/db_engine_specs/trino.py)
+to verify that Superset and Trino have valid interactions.
+
+The main contributions in this pull request was adding a few custom changes for
```
Comment (mosabua) [inline _episodes/12.md:125]: The pull request adds a few basic custom changes to enable Trino usage with Superset.

File: _episodes/12.md [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+like time granularity or SQL syntax, the database engine for that database
+will indicate to Superset that this is not available, and it won't show up as an
+option or report a concise error to the user. By default, database engines use
+the [base.py](https://github.com/apache/superset/blob/master/superset/db_engine_specs/base.py)
+methods, but each engine, like Trino, will add the custom mappings via the
+[trino.py engine](https://github.com/apache/superset/blob/master/superset/db_engine_specs/trino.py)
+to verify that Superset and Trino have valid interactions.
+
+The main contributions in this pull request was adding a few custom changes for 
+Trino to work correctly with Superset. The main change applied was truncating
```
Comment (mosabua) [inline _episodes/12.md:126]: One change ensures that complex timestamps from Trino are truncated to a format

File: _episodes/12.md [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+the complex timestamps coming from Trino to a format that Superset is able to
+support during time aggregation operations.
+
+This along with some tests opened up a vast amount of functionality that we're
```
Comment (mosabua) [inline _episodes/12.md:130]: This opens a fast amount of functionality for using Trino and Superset.

File: _episodes/12.md [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+now able to accomplish with Trino + Superset. We wanted to feature this because
```
Comment (mosabua) [inline _episodes/12.md:131]: with Trino and Superset.

File: _episodes/12.md [hunk 12]
```diff
@@ -0,0 +1,263 @@
...
+it goes to show how a small code change, even one that is not in the Trino
+repository, can have a vast effect on those using Superset and Trino.
+
+Thank you so much to [dungdm93](https://github.com/dungdm93) for making this
+change and further linking Trino into a fantastic project like [Apache
+ Superset](https://superset.apache.org/)!
+
+
+## PR Demo: Superset PR 13105 feat: first step native support Trino
```
Comment (mosabua) [inline _episodes/12.md:140]: Rename as suggested above.

Comment (mosabua) [reply _episodes/12.md:140]: Add info about our guests here maybe .. or in separate section at the top

File: _episodes/12.md [hunk 13]
```diff
@@ -0,0 +1,263 @@
...
+
+To put this PR to the test, we need to connect Superset to Trino as our
+datasource. 
+
+First, you need to follow [these instructions](https://superset.apache.org/docs/installation/installing-superset-using-docker-compose)
+to install Docker (if you don't already have it installed), clone the superset.
```
Comment (mosabua) [inline _episodes/12.md:146]: , and then clone the superset repository:

File: _episodes/12.md [hunk 14]
```diff
@@ -0,0 +1,263 @@
...
+
+`$ git clone https://github.com/apache/superset.git`
+
+Once you have cloned the repository, before you run docker-compose, you want to
```
Comment (mosabua) [inline _episodes/12.md:150]: Next, you need to set up the database driver for Trino. Navigate to the root ...

File: _episodes/12.md [hunk 15]
```diff
@@ -0,0 +1,263 @@
...
+set up the database driver to download. To do this navigate to the root
+directory of the local Superset repository you just downloaded and run the
+following.
+
+`echo "sqlalchemy-trino" >> ./docker/requirements-local.txt`
+
+This tells Superset scripts to install the sqlalchemy-trino library upon
+startup. We know the name by looking up [the Trino driver page](https://superset.apache.org/docs/databases/trino)
+for the driver documentation and how to use the connection string. Note: if you
```
Comment (mosabua) [inline _episodes/12.md:159]: remove the Note:  and just start a normal sentence (maybe rephrased)

File: _episodes/12.md [hunk 16]
```diff
@@ -0,0 +1,263 @@
...
+were to install these directly on a Superset node, you would refer to [this
+database drivers page](https://superset.apache.org/docs/databases/installing-database-drivers).
+
+
+Once this is done run the following to start up Superset and make sure you're in
```
Comment (mosabua) [inline _episodes/12.md:164]: Now run the following command to start up...

File: _episodes/12.md [hunk 17]
```diff
@@ -0,0 +1,263 @@
...
+the root folder of the repo.
+
+`docker-compose -f docker-compose-non-dev.yml up`.
+
+Once Superset is up you will also want to bring up a Trino instance. We did so
```
Comment (mosabua) [inline _episodes/12.md:169]: After Superset is running, you need to start Trino as well.

File: _episodes/12.md [hunk 18]
```diff
@@ -0,0 +1,263 @@
...
+using a seperate docker-compose app. 
+
+As soon as this is done, you can navigate to Superset's homepage <http://localhost:8088>
+and scroll to the Data > Databases menu. 
+
+Click the + Database button.
```
Comment (mosabua) [inline _episodes/12.md:175]: Bold for UI labels and such

```
Click the **+ Database** button
```

File: _episodes/12.md [hunk 19]
```diff
@@ -0,0 +1,263 @@
...
+
+Set the Name to "Trino" and the URI to `trino://trino@host.docker.internal:8080`
```
Comment (mosabua) [inline _episodes/12.md:177]: **Name**

Comment (mosabua) [reply _episodes/12.md:177]: **URI**

File: _episodes/12.md [hunk 20]
```diff
@@ -0,0 +1,263 @@
...
+and click "Add".
```
Comment (mosabua) [inline _episodes/12.md:178]: **Add**

File: _episodes/12.md [hunk 21]
```diff
@@ -0,0 +1,263 @@
...
+
+Note: If you want to allow CTAS, CVAS, or DML operations, you'll want to edit
```
Comment (mosabua) [inline _episodes/12.md:180]: Remove Note: and update

If you want to allow CTAS, CVAS, or DML operations, edit the database, and click on ... 

If

File: _episodes/12.md [hunk 22]
```diff
@@ -0,0 +1,263 @@
...
+the Database you just created and click on the "SQL LAB SETTINGS" tab and check
```
Comment (mosabua) [inline _episodes/12.md:181]: and select the operations you want to allow.

File: _episodes/12.md [hunk 23]
```diff
@@ -0,0 +1,263 @@
...
+in the operations you want to allow.
+<p align="center">
+<img align="center" width="50%" height="100%" src="/assets/episode/12/connection_settings.png"/><br/>
+ Connection settings that allows for creation/manipulation of tables.
+</p>
+
+You should be able to verify under SQL Lab > SQL Editor and run a SELECT query.
```
Comment (mosabua) [inline _episodes/12.md:188]: bold for UI labels.

File: _episodes/12.md [hunk 24]
```diff
@@ -0,0 +1,263 @@
...
+
+We cover adding charts and creating a dashboard in the show but we have linked
```
Comment (mosabua) [inline _episodes/12.md:190]: Break up and simplify sentence

File: _episodes/12.md [hunk 25]
```diff
@@ -0,0 +1,263 @@
...
+a lot of blogs from Preset around how to do a lot of this workflow better than
```
Comment (mosabua) [inline _episodes/12.md:191]: Link to Preset (and maybe say what they do...)

File: _episodes/12.md [hunk 26]
```diff
@@ -0,0 +1,263 @@
...
+even we do during the show which are linked below! Either way, here's a taste
+of what we were able to create in Superset with some [BTS On-Time : Reporting
+Carrier On-Time Performance (1987-present)](https://transtats.bts.gov/Fields.asp?gnoyr_VQ=FGJ)
+and [Covid Cases](https://data.cdc.gov/Case-Surveillance/COVID-19-Case-Surveillance-Public-Use-Data/vbim-akqf) 
+reported by the CDC.
+
+<p align="center">
+<img align="center" width="75%" height="100%" src="/assets/episode/12/covid_flights_data.png"/><br/>
+ Cool Covid Flight data Dashboard we created in Superset!
```
Comment (mosabua) [inline _episodes/12.md:200]: COVID-19 and flights data dashboard

**Review by mosabua — CHANGES_REQUESTED** [review]
Various changes needed .. ping me to reread again when updated

File: _episodes/12.md (hunk 1 above)
Comment (bitsondatadev) [reply _episodes/12.md:3]: I am more in favor of keeping the title short...and we specify Apache Superset in the notes frequently.

File: _episodes/12.md (hunk 3 above)
Comment (bitsondatadev) [reply _episodes/12.md:15]: We would need to go back and change all the rest. Let's leave it for now and I'll make a separate change.

File: _episodes/12.md (hunk 4 above)
Comment (bitsondatadev) [reply _episodes/12.md:16]: This may get long in the summaries but how about.

Demo of Superset querying Trino to create visualizations

File: _episodes/12.md (hunk 3 above)
Comment (bitsondatadev) [reply _episodes/12.md:15]: Actually I decided we can just be inconsistent and i'll still add the separate change later.

File: _episodes/12.md (hunk 5 above)
Comment (bitsondatadev) [reply _episodes/12.md:109]: Agreed. Done

**Review by mosabua — APPROVED** [review]
Looks great now!
